### PR TITLE
Fix workdir and expand arg location.

### LINKED ISSUE #355: nodejs_image does not preserve data filepaths
In nodejs_binary, you can pass in data that will be accessible to the binary at runtime. 

When replaced with nodejs_image, these files are not available when executed. If you manually cd into the workspace directory under the runfiles and run the .binary file via sh ../../{binary_file}, it will work as expected.

Does this have to do with the symlinks in app_layer?

### PR DESCRIPTION
**Pull request by Globegitter**
This supersedes #365. And should fix #355, #378, #451

### CODE REVIEW

**Comment by nlopezgi** [comment]
Thanks a lot for sending this PR! Looks good overall, will review in more detail in a bit, kicked off presubmits to make sure everything is passing.

File: python3/image.bzl [hunk 1]
```diff
@@ -74,6 +74,8 @@ def py3_image(name, base = None, deps = [], layers = [], **kwargs):
     # TODO(mattmoor): Consider using par_binary instead, so that
     # a single target can be used for all three.
 
+    args = kwargs.pop("args", None)
```
Comment (nlopezgi) [inline python3/image.bzl]: why is this necessary for py3 image but not for others?

File: testdata/py3_image.py [hunk 2]
```diff
@@ -19,6 +21,9 @@ def main():
   print('Second: %d' % py3_image_library.fn(2))
   print('Third: %d' % py3_image_library.fn(3))
   print('Fourth: %d' % py3_image_library.fn(4))
+  print(sys.argv)
```
Comment (nlopezgi) [inline testdata/py3_image.py]: can you add short doc line somewhere indicating these two tests (py3, py) expect the third argument to be a file location

File: lang/image.bzl [hunk 3]
```diff
@@ -251,9 +255,10 @@ app_layer = rule(
 
         # Override the defaults.
         "data_path": attr.string(default = "."),
-        "workdir": attr.string(default = "/app"),
+        "workdir": attr.string(default = ""),
         "directory": attr.string(default = "/app"),
         "legacy_run_behavior": attr.bool(default = False),
+        "data": attr.label_list(cfg = "data", allow_files = True),
```
Comment (nlopezgi) [inline lang/image.bzl]: iiuc, this should have as value host/target? https://docs.bazel.build/versions/master/skylark/lib/attr.html#label_list

File: lang/image.bzl [hunk 4]
```diff
@@ -211,21 +212,24 @@ def _app_layer_impl(ctx, runfiles = None, emptyfiles = None):
         _binary_name(ctx): _final_file_path(ctx, ctx.executable.binary),
         # Create a directory symlink from <workspace>/external to the runfiles
         # root, since they may be accessed via either path.
-        _external_dir(ctx): _runfiles_dir(ctx),
+        _external_dir(ctx): runfiles_dir(ctx),
     })
 
+    args = [ctx.expand_location(arg, ctx.attr.data) for arg in ctx.attr.args]
```
Comment (nlopezgi) [inline lang/image.bzl]: add doc comment above explaining you are expanding locations of the form $(location :some_target)

File: java/image.bzl [hunk 5]
```diff
@@ -215,7 +219,9 @@ jar_app_layer = rule(
         "directory": attr.string(default = "/app"),
         # https://github.com/bazelbuild/bazel/issues/2176
         "data_path": attr.string(default = "."),
+        "workdir": attr.string(default = ""),
         "legacy_run_behavior": attr.bool(default = False),
+        "data": attr.label_list(cfg = "data", allow_files = True),
```
Comment (nlopezgi) [inline java/image.bzl]: should be host/target https://docs.bazel.build/versions/master/skylark/lib/attr.html#label_list

File: java/image.bzl [hunk 6]
```diff
@@ -166,12 +168,13 @@ def _jar_app_layer_impl(ctx):
 
     binary_path = layer_file_path(ctx, ctx.files.binary[0])
     classpath_path = layer_file_path(ctx, classpath_file)
+    args = [ctx.expand_location(arg, ctx.attr.data) for arg in ctx.attr.args]
```
Comment (nlopezgi) [inline java/image.bzl]: add doc comment about expanding location

File: container/layer_tools.bzl [hunk 7]
```diff
@@ -158,6 +158,8 @@ def incremental_load(
             ),
         ]
         if run:
+            # bazel automatically passes ctx.attr.args to the binary on run, so need to truncate
+            # them here to not pass twice, as they are already part of the docker entrypoint
```
Comment (nlopezgi) [inline container/layer_tools.bzl]: Can you add a pointer to the Bazel issue you opened so we can track and remove once that is fixed?

File: python3/image.bzl (hunk 1 above)
Comment (Globegitter) [reply python3/image.bzl]: Sorry this is still a leftover, I also tried to fix the issue that args are being passed in twice.

File: container/layer_tools.bzl (hunk 7 above)
Comment (Globegitter) [reply container/layer_tools.bzl]: This is also still a leftover from when I tried to also fix that issue. Yes will clean up.

File: lang/image.bzl (hunk 3 above)
Comment (Globegitter) [reply lang/image.bzl]: I have seen some doc referring that for data it should still be data, there is also a ticket for it: https://github.com/bazelbuild/bazel/issues/4316. But here: https://github.com/bazelbuild/examples/blob/5d1e5bd82e061e8e2e04bb54883fc4c79c935f76/rules/runfiles/execute.bzl#L32 there is nothing, so I will just remove it.
